fs/smb/client/truenas_streams.c

```diff
@@ -0,0 +1,633 @@
+/* SPDX-License-Identifier: LGPL-2.1 */
+/*
+ *
+ *   Copyright (c) iXsystems, inc (2024)
+ *   Author(s): Andrew Walker
+ *
+ */
+
+#include <linux/xattr.h>
+#include "cifspdu.h"
+#include "cifsglob.h"
+#include "cifsproto.h"
+#include "cifs_debug.h"
+#include "truenas_streams.h"
+#include "cifs_unicode.h"
+#include "smb2glob.h"
+#include "smb2pdu.h"
+#include "smb2proto.h"
+
+#define STREAM_MAX_RETRIES 5
+
+unsigned int streams_samba_compat_enabled = 1;
+
+struct parsed_stream_info {
+	u32 size;
+	u32 allocation_size;
+	char *stream_name;
+};
+
+/*
+ * See MS-FSCC 2.4.43 FileStreamInformation
+ *
+ * NOTE regarding stream name:
+ * It is a sequence of unicode characters containing the name of the stream
+ * using the form ":streamname:$DATA" or "::$DATA" for the default data stream.
+ * as specified in MS-FSCC 2.1.4. The field is not null-terminated and MUST
+ * be handled as a sequence of `namelength` bytes.
+ *
+ * NOTE regarding aligment:
+ * When multiple FILE_STREAM_INFORMATION data elements are present in the
+ * buffer, each MUST be alinged on an 8-byte boundaries; any bytes inserted
+ * for alignment SHOULD be set to zero and the receiver MUST ignore them.
+ */
+
+static int
+validate_stream_info(struct smb2_file_stream_info *stream,
+		     char *end)
+{
+	// Check overall stream info
+	char *stream_buf = (char *)stream + sizeof(struct smb2_file_stream_info);
+	if ((stream_buf + stream->StreamNameLength) > end) {
+		return -EOVERFLOW;
+	}
+
+	if (stream->StreamNameLength < DEFAULT_DATA_STREAMLEN) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static struct smb2_file_stream_info
+*get_next_stream(struct smb2_file_stream_info *in,
+		 char *end_of_streams, int *perror)
+{
+	char *stream_buf = (char *)in;
+
+	if (in->NextEntryOffset == 0) {
+		// MS-FSCC 2.4.43 - zero means no more entries
+		return NULL;
+	}
+
+	stream_buf += in->NextEntryOffset;
+	if (stream_buf > end_of_streams) {
+		*perror = -EOVERFLOW;
+		return NULL;
+	}
+
+	return (struct smb2_file_stream_info *)stream_buf;
+}
+
+static int
+parse_to_xat_buf(struct cifs_sb_info *cifs_sb,
+		 char *dst, size_t dstsz,
+		 struct smb2_file_stream_info *in,
+		 char *end_of_streams,
+		 u32 *pused)
+{
+	char *name = NULL, *stream_name;
+	struct smb2_file_stream_info *stream = NULL;
+	int error = 0;
+	u32 used = 0;
+	size_t namelen = 0;
+
+	for (stream = in;
+	     stream != NULL;
+	     stream = get_next_stream(stream, end_of_streams, &error)) {
+		error = validate_stream_info(stream, end_of_streams);
+		if (error) {
+			break;
+		}
+
+		/*
+		 * `name` is UTF-16 and not NULL-terminated and so convert
+		 * to NULL-terminated UTF-8 string
+		 */
+		name = (char *)stream + sizeof(struct smb2_file_stream_info);
+
+		stream_name = cifs_strndup_from_utf16(name, stream->StreamNameLength, true,
+						      cifs_sb->local_nls);
```

Comment:
Would it make sense to check here or in get_next_stream() that `name + stream->StreamNameLength <= end_of_streams`?